File: Assignment1/Question2/Mat_Mul_Double.py
import time
import numpy as np

def matrix_multiply_double(N):
    A = np.ones((N, N), dtype=float)
    B = np.ones((N, N), dtype=float)
    C = np.zeros((N, N), dtype=float)

    start = time.time()

    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i][j] += A[i][k] * B[k][j]

    end = time.time()

    print(f"Time taken for N={N} (double): {end - start:.6f} s")

if __name__ == "__main__":
    matrix_multiply_double(1024) # Hard Code N here for sizes = [64, 128, 256, 512, 1024]

File: Assignment1/Question2/Mat_Mul_Double.cpp
#include <iostream>
#include <vector>
#include <ctime>

using namespace std;

void matrix_multiply_double(int N) {
    vector<vector<double> > A(N, vector<double>(N, 1.0));
    vector<vector<double> > B(N, vector<double>(N, 1.0));
    vector<vector<double> > C(N, vector<double>(N, 0.0));

    clock_t start = clock();

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            for (int k = 0; k < N; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    clock_t end = clock();
    double diff = static_cast<double>(end - start) / CLOCKS_PER_SEC;
    cout << "Time taken by double matrix multiplication: " << diff << " seconds" << endl;
}

int main() {    
    matrix_multiply_double(1024); // Hard Code N here in  [64, 128, 256, 512, 1024]
    return 0;
}
